src/chainparams.cpp

```diff
@@ -353,33 +351,78 @@ class CRegTestParams : public CChainParams {
+        for (const std::string& strDeployment : args.GetArgs("-vbparams")) {
+            std::vector<std::string> vDeploymentParams;
+            boost::split(vDeploymentParams, strDeployment, boost::is_any_of(":"));
+            if (vDeploymentParams.size() != 3) {
+                throw std::runtime_error("Version bits parameters malformed, expecting deployment:start:end");
+            }
+            int64_t nStartTime, nTimeout;
+            if (!ParseInt64(vDeploymentParams[1], &nStartTime)) {
+                throw std::runtime_error(strprintf("Invalid nStartTime (%s)", vDeploymentParams[1]));
+            }
+            if (!ParseInt64(vDeploymentParams[2], &nTimeout)) {
+                throw std::runtime_error(strprintf("Invalid nTimeout (%s)", vDeploymentParams[2]));
+            }
+            bool found = false;
+            for (int j=0; j<(int)Consensus::MAX_VERSION_BITS_DEPLOYMENTS; ++j)
+            {
+                if (vDeploymentParams[0].compare(VersionBitsDeploymentInfo[j].name) == 0) {
+                    UpdateVersionBitsParameters(Consensus::DeploymentPos(j), nStartTime, nTimeout);
+                    found = true;
+                    LogPrintf("Setting version bits activation parameters for %s to start=%ld, timeout=%ld\n", vDeploymentParams[0], nStartTime, nTimeout);
+                    break;
+                }
+            }
+            if (!found) {
+                throw std::runtime_error(strprintf("Invalid deployment (%s)", vDeploymentParams[0]));
+            }
+        }
+    }
+}
+
 static std::unique_ptr<CChainParams> globalChainParams;
 
 const CChainParams &Params() {
     assert(globalChainParams);
     return *globalChainParams;
 }
 
-std::unique_ptr<CChainParams> CreateChainParams(const std::string& chain)
+std::unique_ptr<CChainParams> CreateChainParams(const std::string& chain, ArgsManager& args)
 {
     if (chain == CBaseChainParams::MAIN)
         return std::unique_ptr<CChainParams>(new CMainParams());
     else if (chain == CBaseChainParams::TESTNET)
         return std::unique_ptr<CChainParams>(new CTestNetParams());
     else if (chain == CBaseChainParams::REGTEST)
-        return std::unique_ptr<CChainParams>(new CRegTestParams());
+        return std::unique_ptr<CChainParams>(new CRegTestParams(args));
     throw std::runtime_error(strprintf("%s: Unknown chain %s.", __func__, chain));
 }
 
-void SelectParams(const std::string& network)
+std::unique_ptr<CChainParams> CreateChainParams(const std::string& chain)
 {
-    SelectBaseParams(network);
-    globalChainParams = CreateChainParams(network);
+    ArgsManager args;
+    return CreateChainParams(chain, args);
 }
 
-void UpdateVersionBitsParameters(Consensus::DeploymentPos d, int64_t nStartTime, int64_t nTimeout)
+void SelectParams(const std::string& network)
 {
-    globalChainParams->UpdateVersionBitsParameters(d, nStartTime, nTimeout);
+    SelectBaseParams(network);
+    globalChainParams = CreateChainParams(network, gArgs);
```

Comment:
Why not reference `gArgs` directly in `CreateChainParams()` when using REGTEST? This is the only place that actually calls `CreateChainParams` with the `args` parameter and the only place that calls it with `chain==REGTEST`. That way you don't need to have two different ways of calling `CreateChainParams`.

Reply:
I can remove the new way of calling CreateChainParams, but the idea was that some unittests could create their own custom ArgsManager instead of using gArgs. This would only serve for testing functions that explicitly take a CChainParams instead of relaying on Params().

Reply:
Tend to agree with @ajtowns.

> but the idea was that some unittests could create their own custom ArgsManager instead of using gArgs

If needed it can mutate gArgs? Or this can be done when needed?

Reply:
I guess one can mutate gArgs to a given value for a given test and then back to its original value afterwards.
Or as you say, make this change afterwards when it's going to be used.